sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java

```diff
@@ -449,54 +520,155 @@ static RunQueryRequest makeRequest(Query query, @Nullable String namespace) {
 
     @Override
     public void validate(PBegin input) {
-      checkNotNull(getProjectId(), "projectId");
-      checkNotNull(getQuery(), "query");
+      checkNotNull(getProjectId(), "projectId ValueProvider");
+
+      if (getProjectId().isAccessible()) {
+        checkNotNull(getProjectId().get(), "projectId");
+      }
+
+      if (getQuery() == null && getGqlQuery() == null) {
+        throw new IllegalArgumentException(
+            "Either query or gql query ValueProvider should be provided");
+      }
+
+      if (getQuery() != null && getGqlQuery() != null) {
+        throw new IllegalArgumentException(
+            "Only one of query or gql query ValueProvider should be provided");
```

Comment:
If it's possible, having simply getQuery be the provider and derived from GqlQuery at runtime, as I suggested elsewhere, would probably make this check much easier and satfisfy Ben's use case too.